### ISSUE TITLE
ORDictionary<string, POC> missing items

### ISSUE THREAD

**Issue by tesgu** [akkadotnet/akka.net#4843]
Hi guys

We are experiencing a weird behavior with ORDictionary. Before adding the new item to the dictionary we log the contents of it and, obviously the item doesn't exist. We add the item to the dictionary with WriteLocal consistency and the AddOrUpdate method. Before exiting the Merge function in the update message to the replicator, we log the items in the dictionary and the item exists. Sometimes, when we try to get the item from the dictionary with ReadLocal in the same machine or in the other machines, the item doesn't exist. Nobody has actively remove it, but the item doesn't exist. Item keys are completely unique (guids) and there is only one AddOrUpdate for all the machines... we trust gossip to propagate the value to the other machines.

This happens in a cluster with three machines that are adding and removing items from the dictionary. Not sure what is happening.

As I told we are adding new items with AddOrUpdate, removing them with Remove and getting the items with TryGetValue.

The POC is also derived from IReplicated data. It has a creation date and the Merge method always returns the newest item. If it merges with null, the non null value is returned.

Not sure what is happening... I would think that when trying to get from other machines maybe the gossip didn't work and can not find the value, but when trying to get it in the local machine where the AddOrUpdate was executed... I think it should exist if nobody has deleted it (and nobody did at least actively)

This is our code for the POC merge:
```
        public TransactionState Merge(TransactionState other) =>
            other != null && other.Created > this.Created ? other : this;

        /// <inheritdoc />
        public IReplicatedData Merge(IReplicatedData other) =>
            other is TransactionState state ? this.Merge(state) : this;
```

Any clue about what could be happening?

**Comment by Arkatufus** [comment]
@tesgu I'll be looking into this shortly

**Comment by Arkatufus** [comment]
@tesgu 
Can try to subscribe to the local `Replicator` and check that you never receive any `DataDeleted` message from your `Replicator`?
Can you also turn Debug logging on and check that the number of "Received update for deleted key {0}" messages matches the number of received `DataDeleted` message count?

**Comment by tesgu** [comment]
Sorry @Arkatufus . Maybe we are using distributed data in a way it's not intended to because we are not receiving any DataDeleted message from the replicator as we are not deleting any key. I explain:

- We create an ORDictionaryKey<string, TransactionState> that contains an ORDictionary<string, TransactionState>.
- When we wat to add a TransactionState to the ORDictionary we use AddOrUpdate method in the Merge method passed in the Update message for the replicator
```
            var value = new TransactionState(message.TransactionId, message.Created);

            var updateCommand = new Update(
                new ORDictionaryKey<string, TransactionState>("transaction-states"),
                ORDictionary<string, TransactionState>.Empty,
                WriteLocal.Instance,
                oldValue =>
                {
                    var oldDictionary = oldValue.AsInstanceOf<ORDictionary<string, TransactionState>>();
                    return oldDictionary.AddOrUpdate(this.cluster, value.TransactionId, null, oldTrxState => value.Merge((TransactionState)oldTrxState));
                });

            this.replicator.Tell(updateCommand);
 ````
- When we want to remove a TransactionState from the ORDictionary, we use the method Remove in the merge method passed to the Update message for the replicator.
```
        private void RemoveItem(TransactionState result) =>
             this.replicator.Tell(
                new Update(
                    new ORDictionaryKey<string, TransactionState>("transaction-states"),
                    ORDictionary<string, TransactionState>.Empty,
                    WriteLocal.Instance,
                    oldValue =>
                    {
                        var oldDictionary = oldValue.AsInstanceOf<ORDictionary<string, TransactionState>>();
                        var newDictionary = oldDictionary.Remove(this.cluster.SelfUniqueAddress, result.TransactionId);
                        TransactionStorageLogEvent
                            .TransactionRemoved(result.TransactionId)
                            .Log(this.logger);

                        return newDictionary;
                    }));
 ```
- When we want to get a TransactionState from the ORDictionary, we use the method TryGetValue on the ORDictionary returned by the Get message.
```
                var getDictionaryResponse = await this.replicator.Ask<IGetResponse>(
                    new Get(new ORDictionaryKey<string, TransactionState>("transaction-states"), ReadLocal.Instance),
                    TimeSpan.FromSeconds(this.clusterSettings.OperationTimeOut));
                    
                switch (getDictionaryResponse)
                {
                    case GetSuccess getSuccess:
                        var getDictionary = getSuccess.Data.AsInstanceOf<ORDictionary<string, TransactionState>>();
                        return getDictionary.TryGetValue(transactionId, out var trxState) ? trxState : null;
 ```
 I suppose (is not specified in the docs) that the DataDeleted event is sent, for example when you delete via the Replicator Delete message a key, in our case the ORDictionaryKey. I say that, because I am not receiving any DataDeleted message from any for the replicators, I only receive Changed messages.

Is this the way ORDictionary is intended to be used? Our problem comes in the Get part. Although we are sure that the TransactionState exists in the ORDictionary three machines, the Get doesn't find it in the returned dictionary.

Maybe there is a problem with the gossip and the ORDictionary?

**Comment by Arkatufus** [comment]
So when you said that the item does not exist, do you mean the ORDictionary key itself was missing, or was it the value of the ORDictionary key that failed to update, propagate or somehow got deleted?

**Comment by tesgu** [comment]
A value in the ORDictionary. When we insert is present instantly in the local machine and via gossip in the remote machines. We know that because we are printing the dictionary contents in the logs. But when we perform the get it doesn't appear... Somehow 'magically' it's not there...

**Comment by Arkatufus** [comment]
@tesgu I was just confused about what you ment by missing, I thought that you were talking about keys not propagated across, I get the problem now and i am still looking into it.

**Comment by Horusiath** [comment]
@tesgu

> We add the item to the dictionary with WriteLocal consistency and the AddOrUpdate method. [...] Sometimes, when we try to get the item from the dictionary with ReadLocal in the same machine or in the other machines, the item doesn't exist.

Are we still talking about the same machine here? I mean, if you did write followed by read on the same machine, you should always receive read response with an updated state. If it's not it means that the update has failed for some reason - most probably modify function passed with update has thrown an exception or perhaps the object itself couldn't be serialized -  and value was never written.

**Comment by tesgu** [comment]
If we add the item and immediately after that perform the read, the item exists in the local machine. After a while, other inserts and deletes had happened in the local machine and the remote ones, if we try to read from the local machine, sometimes, the value is not there. Same with remote machines: We update local machine, value exists in local machine, value exists in remote machines (via gossip), after a while, after other removal and inserts had been done, when we try to read on a remote machine the value, sometimes, it doesn't exist.

I think this weird behavior is going to be difficult to reproduce, becuase it doesn't happen always... We are not able to launch our product using DData in production due to this behavior... My team and my boss has agreed to desestimate Akka as a possible solution to our problem, so if you want to, guys, you can close the issue. 

Sad, because we have invested a lot of time implementing the solution using Akka and we really though that Akka is going to resolve the problem and maybe in a future use it for another products in the company. I really like the concept behind Akka, but there has been some bugs with DData and now my boss doesn't trust much in the framework, as ours is a critical product.

Sorry for the inconveniences

**Comment by Aaronontheweb** [comment]
> I think this weird behavior is going to be difficult to reproduce, becuase it doesn't happen always... We are not able to launch our product using DData in production due to this behavior... My team and my boss have agreed to desestimate Akka as a possible solution to our problem, so if you want to, guys, you can close the issue.

@Arkatufus is looking into it and fixing it, as we want to make sure the library performs as advertised and expected. DData has been out of beta for a little less than a year and it's inevitable that we're going to run into some issues like this - it's true of any software product, doubly so for something as complex as distributed in-memory CRDT-based replication. 

Your team and your boss should be pleased that we've been responsive to your inquiries and have promptly fixed all of the issues you've reported at no cost thus far. They're not going to find that level of responsiveness or support without paying handsomely for it elsewhere. I know that's not your decision - your bug reports and contributions to the project have been _very_ welcome and have helped us make DData much better - but you might want to reiterate that to your team.

**Comment by tesgu** [comment]
@Aaronontheweb don't take me wrong. I am amazed with your responsiveness and thanks to you all previous issues has been solved quickly. In fact I was amazed that for a Open Source project being free your reponse has always being quick. But you know... my boss doesn't think as me, and we have a deadline...

**Comment by Aaronontheweb** [comment]
@tesgu no worries - I get it. I know it's not you. But what I am saying, as a CTO with deadlines myself, is I think your boss' expectations that you're going to find a better experience out there at lower cost is going to result in disappointment.

**Comment by tesgu** [comment]
Totally agree @Aaronontheweb

**Comment by Aaronontheweb** [comment]
Resolved via #4910 - we think
